src/charm.py

```diff
@@ -122,77 +145,82 @@ def on_agent_relation_joined(self, event):
         event.relation.data[self.model.unit]["labels"] = labels
         event.relation.data[self.model.unit]["slavehost"] = agent_name
 
-    def on_agent_relation_changed(self, event):
+    def _on_agent_relation_changed(self, event: charm.RelationChangedEvent):
         """Populate local configuration with data from relation."""
         logger.info("Jenkins relation changed")
+        agent_name = self._gen_agent_name()
+        self._stored.agents = [agent_name]
+        self._stored.agent_tokens = self._stored.agent_tokens or []
+
+        # Check event data
         try:
-            self._stored.jenkins_url = event.relation.data[event.unit]['url']
+            relation_jenkins_url = event.relation.data[event.unit]['url']
         except KeyError:
-            pass
-
+            logger.warning(
+                f"Expected 'url' key for {event.unit} unit in relation data. "
+                "Skipping setup for now."
+            )
```

Comment:
Same comment as above about logging and building strings.

Reply:
Yes, I actually had fixed this in the next PR

Reply:
Fixed